fabfile/add.py

```diff
@@ -2,119 +2,230 @@
+        utils.psql("DROP DATABASE IF EXISTS contrib_regression WITH (FORCE);")
+        utils.psql("DROP DATABASE IF EXISTS regression WITH (FORCE);")
+
+        # create result folder if not exists
+        utils.mkdir_if_not_exists(config.RESULTS_DIRECTORY)
+
+        # load depended extensions
+        self.create_depended_extensions()
+
+        # run tests with warn_only option to not exit in case of a test failure in the extension,
+        # because we want to run regression tests for other extensions too.
+        with settings(warn_only=True):
+            with cd(self.test_path):
+                run(self.test_command)
+
+        # rename regression.diffs, if exists, so that extension's diff file does not conflict with others'        
+        self.rename_regression_diff()
+
+    def rename_regression_diff(self):
+        regression_diffs_path = os.path.join(self.test_path, config.REGRESSION_DIFFS_FILE)
+        if os.path.isfile(regression_diffs_path):
+            extension_regression_diff = config.REGRESSION_DIFFS_FILE + '_{}_'.format(self.test_name) + self.extension.name
+            with cd(config.RESULTS_DIRECTORY):
+                run('mv {} {}'.format(regression_diffs_path, extension_regression_diff))
+
+
+class InstallExtensionTask(Task):
+    '''
+    A class which has all the boilerplate for building and installing extensions.
+    '''
+
+    def __init__(self, extension, **kwargs):
+        self.name = extension.name  # the name of the task (fab [xxx])
+        self.__doc__ = extension.doc  # the description which fab --list will list
+        self.contrib = extension.contrib
+        self.repo_url = extension.repo_url
+        self.default_git_ref = extension.default_git_ref
+        self.preload = extension.preload
+        self.create = extension.create
+        self.configure = extension.configure
         self.before_run_hook = kwargs.get('before_run_hook', None)
         self.post_install_hook = kwargs.get('post_install_hook', None)
 
         super(InstallExtensionTask, self).__init__()
 
-    @staticmethod
-    def repo_path_for_url(repo_url):
-        repo_name = run('basename {}'.format(repo_url))
-        repo_name = repo_name.split('.')[0] # chop off the '.git' at the end
-        return os.path.join(config.CODE_DIRECTORY, repo_name)
+    def create_extension(self):
+        if self.create:
+            utils.psql('CREATE EXTENSION {};'.format(self.name))
 
     def run(self, *args):
         if self.before_run_hook:
             self.before_run_hook()
 
-        prefix.check_for_pg_latest()  # make sure we're pointed at a real instance
-        utils.add_github_to_known_hosts() # make sure ssh doesn't prompt
+        extension_path = ''
+        if self.contrib:
+            extension_path = contrib_path_for_extension(self.name)
+        else:
+            extension_path = extension_path_for_url(self.repo_url)
```

Comment:
Because we need connection object given in the first argument of extension_path_for_url(c, self.repo_url). Constructor cannot be given the connection.